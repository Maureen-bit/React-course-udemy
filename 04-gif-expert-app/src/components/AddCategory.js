import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({setCategories}) => {
    const [inputValue, setInputValue] = useState('');
    const changeText = (event) => {
        setInputValue(event.target.value);
    };

    const onSubmitForm = (event) =>{
        /* Evitar comportamiento pot defecto del formulario de cambiar de ruta 
        ( de http://localhost:3000 a http://localhost:3000/?) y hacer post de la informacion del formulario  */
        event.preventDefault();
        if(inputValue.trim().length > 2){
        /* Se llama como un call back para tener como referencia el estado anterior, es decir el valor
        de categories en el useState */
        setCategories((categories) => [ inputValue, ...categories ]);
        /* Para que al dar enter limpie el input y quede vacio para escribir otra cosa */
        setInputValue('');
        console.log('Submit hecho!')
    }
    }

  return (
      <form onSubmit={(event) => onSubmitForm(event)} >
        <p>{inputValue}</p>
        <input 
            type="text"
            value={inputValue}
            onChange={ (event) => changeText(event)}
        />
    </form>
  )
};

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
  };