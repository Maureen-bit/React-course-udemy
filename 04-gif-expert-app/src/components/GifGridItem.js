import React from 'react';
import PropTypes from 'prop-types';
import 'animate.css';

export const GifGridItem = ( {title, url} ) => {
  return (
    <div className='card animate__animated animate__heartBeat'>
        <img src={url} alt={title}/>
        <p>{title}</p>
    </div>
  )
};

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired
};