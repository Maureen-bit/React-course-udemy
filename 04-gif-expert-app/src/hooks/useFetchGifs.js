import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGIF";

export const useFetchGifts = (category ) => {

    /* Custom hooks son diferentes a functional components, el custom hook se 
    encarga de hacer el trabajo pesado, de la logica fuerte, como obtener informaciÃ³n de las Api y retornarlas al 
    functional component para que las muestre */
    const [state, setstate] = useState({
        data: [],
        loading: true
    });

    useEffect(()=>{
        getGifs(category)
            .then( images => 
                setstate({
                    data: images,
                    loading: false
                })
            )
    }, [category]);

    return state;
}