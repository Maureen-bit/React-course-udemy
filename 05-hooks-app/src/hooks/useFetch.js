import { useEffect, useRef, useState } from 'react';

export const useFetch = ( url ) => {
    /* Voy a guardar la referencia  de cuando se monta el componente, como se renderiza por primera vez
    entonces este valor será true*/
    const isMounted = useRef(true);
    const [state, setState] = useState({data: null, loading: true, error: null});

    useEffect(() => {    
      return () => {
          /* Este return se disparará cada vez que se desmonte el componente por eso cambiamos el valor de la
          referencia a false */
        isMounted.current = false;
      };
    }, []);
    
    useEffect(() => {
        setState({data: null, loading: true, error: null});
        fetch(url)
        .then(resp => resp.json())
        .then(data => {
            /* Usamos el valor de la referencia para ejecutar el setState cuando el componente
            esté montado y evitar el error de: 
            Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory 
            leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function. */
            if (isMounted.current){
                /* Si está en true es porque el componente está montado */
                setState({
                    loading: false,
                    error: null,
                    data
                })
            } else {
                console.log("SetState no se llamó")
            }
        })
    }, [url]);

    return state;
};
