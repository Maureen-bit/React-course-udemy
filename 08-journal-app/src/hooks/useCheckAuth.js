import { onAuthStateChanged } from 'firebase/auth';
import { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { FirebaseAuth } from '../firebase/config';
import { login, logout } from '../store/auth';

export const useCheckAuth = () => {
    const { status } = useSelector(state => state.auth);
    const dispatch = useDispatch();
    
    useEffect(() => {
      /** onAuthStateChanged: función que está pendiente de los cambios en el estado de la autenticación del usuario */
      onAuthStateChanged(FirebaseAuth, async (user) => {
        if(!user){
          return dispatch(logout());
        }
  
        const { uid, email, displayName, photoURL } = user;
        dispatch(login({uid, email, displayName, photoURL}))
  
      });
  
    }, [])
  return { status }
}
