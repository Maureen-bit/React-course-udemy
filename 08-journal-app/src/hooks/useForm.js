import { useEffect, useMemo, useState } from 'react';

export const useForm = ( initialForm = {}, formValidations = {} ) => {
  
    const [ formState, setFormState ] = useState( initialForm );
    /** Cuando requerimos volver a renderizar por cambios en el estado debemos usar un hook que maneje el estado */
    const [ formValidation, setFormValidation ] = useState( {} );

    useEffect(() => {
        createValidators();
    }, [formState]);

    const isFormValid = useMemo(() =>{
        for (const formValue of Object.keys(formValidation)) {
            if( formValidation[formValue] !== null) {
                return false;
            }
        }
        return true;
    }, [formValidation]);

    const onInputChange = ({ target }) => {
        const { name, value } = target;
        setFormState({
            ...formState,
            [ name ]: value
        });
    }

    const onResetForm = () => {
        setFormState( initialForm );
    }

    const createValidators = () => {
        const formCheckedValues = {};
        for (const formField of Object.keys(formValidations)) {
            /**  Devuelve cada arreglo de cada propiedad (email, password, displayName)*/
            const [ funcionValidacion, errorMessage = 'Este valor es requerido' ] = formValidations[formField];
            /** creación de variable dinámica dentro del arreglo para cada propiedad (email, password, displayName), 
             * según resultado de la función, si es válido no devuelve nada porque no hay error pero si hay error devuelve el mensaje
             * de error.
             */
            formCheckedValues[`${ formField}Valid`] = funcionValidacion(formState[formField]) ? null : errorMessage;
        }
        setFormValidation(formCheckedValues);
    }



    return {
        ...formState,
        formState,
        onInputChange,
        onResetForm,
        ...formValidation,
        isFormValid
    }
}