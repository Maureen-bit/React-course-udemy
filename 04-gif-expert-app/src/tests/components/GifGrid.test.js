import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifts } from "../../hooks/useFetchGifs";
/* Creación de mock para fingir/simular la llamada a este archivo y suponer o controlar la información que 
este useFetch va a responder*/
jest.mock('../../hooks/useFetchGifs');

describe("Pruebas unitarias componente <GifGrid />", () => {
    const category = "Heidi";
    test('Debe de mostrar el componente correctamente', () => {
        /* Cuando se llame en el componente retornará el valor */
        useFetchGifts.mockReturnValue(
            {
                /*Aqui dentro va el resultado que yo esperaría del useFetch, iniciamos con los valores que el componente tiene por defecto
                para renderizarse por primera vez, es decir, cuando no hay data aún */
                data: [],
                loading: true
            }
        );
        const view = shallow(<GifGrid category={category} />);
        expect(view).toMatchSnapshot();
    });    
    test('Debe de mostrar items cuando se cargan imagenes después de useFetchGifs', () => {
        const gifs = [{
            id: 'ABC123',
            title: 'The Simpsons Family',
            url: 'https://www.giftestthasimpsonsfamily.com'
        }]
        /* Simulación de que ya cargó data */
        useFetchGifts.mockReturnValue({
                data: gifs,
                loading: false
        });
        const view = shallow(<GifGrid category={category} />);
        /* Evaluación de existencia o no de elementos */
        /* 1. Como loading está en false entonces el elemento p no debe existir */ 
        /* Nota: en este caso el snapshot no es tan confiable ya que al actualizarlo puede pasar la prueba a pesar
        de que no haya data, por eso es mejor evaluar la existencia de los otros elementos */
        expect(view).toMatchSnapshot();
        expect(view.find('p').exists()).toBe(false);
        expect(view.find('GifGridItem').length).toBeGreaterThan(0);
    });
})