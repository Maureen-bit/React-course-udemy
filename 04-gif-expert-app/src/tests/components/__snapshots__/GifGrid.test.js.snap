// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas unitarias componente <GifGrid /> Debe de mostrar el componente correctamente 1`] = `
<Fragment>
  <h3>
    Heidi
  </h3>
  <p>
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Pruebas unitarias componente <GifGrid /> Debe de mostrar items cuando se cargan imagenes despu√©s de useFetchGifs 1`] = `
<Fragment>
  <h3>
    Heidi
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC123"
      key="ABC123"
      title="The Simpsons Family"
      url="https://www.giftestthasimpsonsfamily.com"
    />
  </div>
</Fragment>
`;
